sunkit_image/offlimb_enhance.py

```diff
@@ -295,3 +297,162 @@ def normalizing_radial_gradient_filter(
+    nbins = radial_bin_edges.shape[1]
+
+    # Storage for the filtered data
+    data = np.zeros_like(smap.data)
+
+    # Iterate over each circular ring
+    for i in range(0, nbins):
+
+        # Finding the pixels which belong to a certain circular ring
+        annulus = np.logical_and(map_r > radial_bin_edges[0, i], map_r < radial_bin_edges[1, i])
+        annulus = np.logical_and(annulus, map_r > application_radius)
+
+        # The angle subtended by each segment
+        segment_angle = 2*np.pi / number_angular_segments
+
+        # Storage of mean and standard deviation of each segment in a circular ring
+        average_segments = np.zeros((1, number_angular_segments))
+        std_dev = np.zeros((1, number_angular_segments))
+
+        # Calculating sin and cos of the angles to be multiplied with the means and standard
+        # deviations to give the fourier approximation
+        cos_matrix = np.cos(np.array([[(2 * np.pi * (j + 1) * (i + 0.5)) / number_angular_segments for j in range(order)] for i in range(number_angular_segments)]))
+        sin_matrix = np.sin(np.array([[(2 * np.pi * (j + 1) * (i + 0.5)) / number_angular_segments for j in range(order)] for i in range(number_angular_segments)]))
+
+        # Iterate over each segment in a circular ring
+        for j in range(0, number_angular_segments, 1):
+       
+            # Finding all the pixels whose angle values lie in the segment
+            angular_segment = np.logical_and(angles > segment_angle * j, angles < segment_angle * (j + 1))
+
+            # Finding the particular segment in the circular ring
+            annulus_segment = np.logical_and(annulus, angular_segment)
+
+            # Finding mean and standard deviation in each segnment. If takes care of the empty slices.
+            if np.sum([annulus_segment > 0]) == 0:
+                average_segments[0, j] = 0
+                std_dev[0, j] = 0
+            else:
+                average_segments[0, j] = intensity_summary(smap.data[annulus_segment])
+                std_dev[0, j] = width_function(smap.data[annulus_segment])
+
+        # Calculating the fourier coefficients multiplied with the attenuation coefficients
+        fourier_coefficient_a_0 = np.sum(average_segments) * (2 / number_angular_segments) * (attenuation_coefficients[0, 1])
+        fourier_coefficients_a_k = np.matmul(average_segments, cos_matrix) * (2 / number_angular_segments) * attenuation_coefficients[0][1:]
+        fourier_coefficients_b_k = np.matmul(average_segments, sin_matrix) * (2 / number_angular_segments) * attenuation_coefficients[0][1:]
+        fourier_coefficient_c_0 = np.sum(std_dev) * (2 / number_angular_segments) * attenuation_coefficients[1, 1]
+        fourier_coefficients_c_k = np.matmul(std_dev, cos_matrix) * (2 / number_angular_segments) * attenuation_coefficients[1][1:]
+        fourier_coefficients_d_k = np.matmul(std_dev, sin_matrix) * (2 / number_angular_segments) * attenuation_coefficients[1][1:]
```

Comment:
Is there no way to calculate the coefficients in bulk?

Reply:
I can squash some matrix multiplications into one by concatenation. But the number of operations would still be the same.

Reply:
That sounds like that could be ugly.

Reply:
Yes. Keeping a track of operations would be difficult.